### ISSUE TITLE
java.util.ConcurrentModificationException

### ISSUE THREAD

**Issue by mourjan** [chat-sdk/chat-sdk-android#199]
```
java.util.ConcurrentModificationException
                                                                             at java.util.HashMap$HashIterator.nextEntry(HashMap.java:788)
                                                                             at java.util.HashMap$KeyIterator.next(HashMap.java:815)
                                                                             at co.chatsdk.ui.profile.EditProfileActivity.saveAndExit(EditProfileActivity.java:271)
                                                                             at co.chatsdk.ui.profile.EditProfileActivity.onOptionsItemSelected(EditProfileActivity.java:215)
                                                                             at android.app.Activity.onMenuItemSelected(Activity.java:3205)
                                                                             at android.support.v4.app.FragmentActivity.onMenuItemSelected(FragmentActivity.java:361)
                                                                             at android.support.v7.app.AppCompatActivity.onMenuItemSelected(AppCompatActivity.java:195)
                                                                             at android.support.v7.view.WindowCallbackWrapper.onMenuItemSelected(WindowCallbackWrapper.java:108)
                                                                             at android.support.v7.app.AppCompatDelegateImplV9.onMenuItemSelected(AppCompatDelegateImplV9.java:674)
                                                                             at android.support.v7.view.menu.MenuBuilder.dispatchMenuItemSelected(MenuBuilder.java:822)
                                                                             at android.support.v7.view.menu.MenuItemImpl.invoke(MenuItemImpl.java:171)
                                                                             at android.support.v7.view.menu.MenuBuilder.performItemAction(MenuBuilder.java:973)
                                                                             at android.support.v7.view.menu.MenuBuilder.performItemAction(MenuBuilder.java:963)
                                                                             at android.support.v7.widget.ActionMenuView.invokeItem(ActionMenuView.java:624)
                                                                             at android.support.v7.view.menu.ActionMenuItemView.onClick(ActionMenuItemView.java:150)
                                                                             at android.view.View.performClick(View.java:5698)
                                                                             at android.widget.TextView.performClick(TextView.java:10903)
                                                                             at android.view.View$PerformClick.run(View.java:22570)
                                                                             at android.os.Handler.handleCallback(Handler.java:739)
                                                                             at android.os.Handler.dispatchMessage(Handler.java:95)
                                                                             at android.os.Looper.loop(Looper.java:148)
                                                                             at android.app.ActivityThread.main(ActivityThread.java:7223)
                                                                             at java.lang.reflect.Method.invoke(Native Method)
                                                                             at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1230)
                                                                             at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1120)
```

**Comment by bensmiley** [comment]
Could you provide some context? What steps need to be taken to reproduce this error?

**Comment by mourjan** [comment]
more from firebase Crashylitics

Fatal Exception: java.util.ConcurrentModificationException

co.chatsdk.ui.profile.EditProfileActivity.saveAndExit (EditProfileActivity.java:271)
co.chatsdk.ui.profile.EditProfileActivity.onOptionsItemSelected (EditProfileActivity.java:215)
android.app.Activity.onMenuItemSelected (Activity.java:3205)
android.support.v4.app.FragmentActivity.onMenuItemSelected (FragmentActivity.java:361)
android.support.v7.app.AppCompatActivity.onMenuItemSelected (AppCompatActivity.java:195)
android.support.v7.view.WindowCallbackWrapper.onMenuItemSelected (WindowCallbackWrapper.java:108)
android.support.v7.app.AppCompatDelegateImplV9.onMenuItemSelected (AppCompatDelegateImplV9.java:674)
android.support.v7.view.menu.MenuBuilder.dispatchMenuItemSelected (MenuBuilder.java:822)
android.support.v7.view.menu.MenuItemImpl.invoke (MenuItemImpl.java:171)
android.support.v7.view.menu.MenuBuilder.performItemAction (MenuBuilder.java:973)
android.support.v7.view.menu.MenuBuilder.performItemAction (MenuBuilder.java:963)
android.support.v7.widget.ActionMenuView.invokeItem (ActionMenuView.java:624)
android.support.v7.view.menu.ActionMenuItemView.onClick (ActionMenuItemView.java:150)
android.view.View.performClick (View.java:5698)
android.widget.TextView.performClick (TextView.java:10903)
android.view.View$PerformClick.run (View.java:22570)
android.os.Handler.handleCallback (Handler.java:739)
android.os.Handler.dispatchMessage (Handler.java:95)
android.os.Looper.loop (Looper.java:148)
android.app.ActivityThread.main (ActivityThread.java:7223)
java.lang.reflect.Method.invoke (Method.java)
com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run (ZygoteInit.java:1230)
com.android.internal.os.ZygoteInit.main (ZygoteInit.java:1120)

**Comment by mourjan** [comment]
@bensmiley I was just trying to test and update the username/nickname.. first I changed the profile ppic and right after the picture changed.. I changed the username.. and it crashed

**Comment by bensmiley** [comment]
@mourjan Thanks for the extra details. Can you confirm that you are using the latest version of the project from Gradle? (4.0.6)

**Comment by mourjan** [comment]
@bensmiley yes, using 4.0.6, just started working with.. I tried to reproduce without any luck.. will try again with a fresh user/install

**Comment by bensmiley** [comment]
This issue was reported recently and I thought I fixed it. Can you let me know if you manage to reproduce it again?

**Comment by mourjan** [comment]
Hi again, as promised I tried with a freshly authenticated user and was able to replicate, so I will explain a step by step: 

(this only happens on the first run of the steps listed below, so to reproduce a new user has to be created)

1- I have implemented a custom firebase token based authentication and after being connected to firebase
2- I run the following "InterfaceManager.shared().a.startLoginActivity(this, true);"
3- chat-sdk auto logs in of course
4- in profile settings I first click change the image
5- right after, I change the name and press done
6- and the crash happens

hope this helps,

if you have time and kind enough to advise me about contact search not returning any results would be really helpful:
1- I tried the default app search activity, without any luck (very sure of user name which it is clearly found in database searchindex table)
2- I also tried the NM.search() with different Keys and also without any luck

am I missing something?

Thank you in advance

**Comment by bensmiley** [comment]
This has been fixed and will be available in version number `4.0.7`.
